### PR TITLE
Regression test for missing frames after exporting a CVAT dataset

### PR DESCRIPTION
**Pull request by archibald1418**
This pull request provides test coverage for changes made in #8662 which helped resolve the issue GH-8483

### Motivation and context

- This addresses lack of frames when exporting a dataset in CVAT format with a non-default frame step (i.e not 1)
- The PR adds tests for 'Export dataset' feature to avoid regressions in the future releases
- Also, _server REST API tests_ debug configuration from `.vscode/launch.json` is patched with extra `pytest` options for easier debugging

### How has this been tested?

The test checks that the exported dataset:
- contains the same number of frames that were originally present in the source data 
- contains the correct frames given the frame step

This testing mainly focuses on export in **CVAT for video** format but also confirms the correct behaviour when the format is different (**COCO** and **Datumaro** were checked)

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

## Summary by CodeRabbit

- **New Features**
	- Added a new test for exporting tasks with non-default frame steps, enhancing validation of export functionality.
	- Improved handling of annotations during import/export processes.

- **Bug Fixes**
	- Updated test suite to ensure compatibility with new functionality and handle edge cases.

- **Chores**
	- Added a comment in the requirements file to indicate a future update for `pytest` and `pytest-timeout`. 

- **Configuration Updates**
	- Enhanced debugging configuration for REST API tests with additional arguments for improved output.

### CODE REVIEW

File: tests/python/rest_api/test_tasks.py [hunk 1]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+        "format_name, num_frames, frame_step, expected_frames",
+        [
+            ("Datumaro 1.0", 100, 5, 20),
+            ("COCO 1.0", 100, 5, 20),
+            ("CVAT for video 1.1", 100, 5, 20),  # no remainder
+            ("CVAT for video 1.1", 97, 5, 20),  # prime
+            ("CVAT for video 1.1", 97, 2, 49),
+            ("CVAT for video 1.1", 100, 3, 34),  # three
+            # we assert that expected frames are ceil(frames / step)
+        ],
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1027]: Why there is no "CVAT for images" in the list?

File: tests/python/rest_api/test_tasks.py [hunk 2]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+    def test_export_with_non_default_frame_step(
+        self,
+        # fixtures
+        tmp_path: Path,
+        admin_user: str,
+        # params
+        format_name: str,
+        num_frames: int,
+        frame_step: int,
+        expected_frames: int,
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1038]:
```suggestion
        expected_frame_count: int,
```
I think, it might be useful to check the actual frame names as well.

File: tests/python/rest_api/test_tasks.py [hunk 3]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+        assert expected_frames == math.ceil(num_frames / frame_step), "Test params are wrong"
+
+        spec = {
+            "name": f"test_video_frames_in_{format_name}_after_export",
+            "labels": [{"name": "goofy ahh car"}],
+        }
+
+        data = {
+            "image_quality": 70,
+            "client_files": [generate_video_file(num_frames)],
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1050]: Shouldn't image tasks be checked as well?

File: tests/python/rest_api/test_tasks.py [hunk 4]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+        with make_sdk_client(admin_user) as client:
+            task_obj: Task = client.tasks.retrieve(task_id)
+
+        # export the video
+        dataset_file = tmp_path / "dataset.zip"
+        task_obj.export_dataset(format_name, dataset_file, include_images=True)
+
+        def get_png_index(zinfo: zipfile.ZipInfo) -> int:
+            name = PurePosixPath(zinfo.filename)
+            if name.suffix.lower() != ".png":  # png is usually for video
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1065]:
```suggestion
```
I don't think it makes sense to add some branching here in the test. There is only 1 expected file format.

File: tests/python/rest_api/test_tasks.py [hunk 5]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+                return -1
+            name = os.path.basename(name).removesuffix(name.suffix)
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1067]:
```suggestion
            name = name.stem
```

File: tests/python/rest_api/test_tasks.py [hunk 6]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+            idx = name[name.rfind("_") + 1 :]
+            assert idx.isnumeric()
+            return int(idx)
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1070]:
```suggestion
            return int(name.stem.rsplit("_", maxsplit=1)[-1])
```

File: tests/python/rest_api/test_tasks.py (hunk 1 above)
Comment (archibald1418) [reply tests/python/rest_api/test_tasks.py:1027]: The original bug was with video
Is it necessary to check for the images too?

File: tests/python/rest_api/test_tasks.py [hunk 7]
```diff
@@ -1012,6 +1012,71 @@ def test_datumaro_export_without_annotations_includes_image_info(
...
+            ("CVAT for video 1.1", 100, 3, 34),  # three
+            # we assert that expected frames are ceil(frames / step)
+        ],
+    )
+    def test_export_with_non_default_frame_step(
+        self,
+        # fixtures
+        tmp_path: Path,
+        admin_user: str,
+        # params
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1034]: Consider adding `start_frame` as well.

File: tests/python/rest_api/test_tasks.py (hunk 3 above)
Comment (archibald1418) [reply tests/python/rest_api/test_tasks.py:1050]: From what I understand, images and videos are processed with the same code, the issue was with the format only
I can add a parametrized test for images, too, if it's necessary

File: tests/python/rest_api/test_tasks.py [hunk 8]
```diff
@@ -6349,3 +6349,72 @@ def check_element_outside_count(track_idx, element_idx, expected_count):
...
+
+        if media_type == _SourceDataType.images:
+            (spec, task_id) = next(self._uploaded_images_task_fxt_base(**args))
+        else:
+            (spec, task_id) = next(self._uploaded_video_task_fxt_base(**args))
+
+        with make_sdk_client(self._USERNAME) as client:
+            task = client.tasks.retrieve(task_id)
+
+        yield (spec, task, f"CVAT for {media_type} 1.1")
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:6383]:
```suggestion
            yield (spec, task, f"CVAT for {media_type} 1.1")
```

Later there is a call to `get_meta()`, it uses the client.

Comment (archibald1418) [reply tests/python/rest_api/test_tasks.py:6383]: You mean, it's better to have it inside the fixture as well?

Comment (zhiltsov-max) [reply tests/python/rest_api/test_tasks.py:6383]: The client is created inside the fixture, but it's context manager is closed before the fixture yields the value. If you want to use the client (saved in the `task` object), don't close the context manager before `yield`.

File: tests/python/rest_api/test_tasks.py [hunk 9]
```diff
@@ -6349,3 +6349,72 @@ def check_element_outside_count(track_idx, element_idx, expected_count):
...
+
+        task_meta = task.get_meta()
+        (src_start_frame, src_stop_frame, src_frame_step) = (
+            task_meta["start_frame"],
+            task_meta["stop_frame"],
+            spec.frame_step,
+        )
+        src_end_frame = (
+            src_stop_frame - ((src_stop_frame - src_start_frame) % src_frame_step) + src_frame_step
+        )  # taken from TestTaskData._compute_annotation_segment_params
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:6415]: Please extract a function for this. The comment in the code would be fine if it was some function from another project, which you can't easily use a dependency, but here it is in the same file.

**Review by zhiltsov-max — APPROVED** [review]

**Comment by Marishka17** [comment]
@archibald1418, could you please next time merge PRs with squash to make the develop branch history more clear?

**Comment by archibald1418** [comment]
TODO: factor out inheritance since it slows down test execution
